File: backend/src/services/messageQueueService.ts
import prisma from '../config/database';
import { logger } from '../utils/logger';
import { smsService } from './smsService';
import { config } from '../config';
import { MessageStatus, QueueStatus } from '@prisma/client';

interface QueuedMessage {
  messageId: string;
  phoneNumber: string;
  content: string;
  clientId: string;
}

class MessageQueueService {
  private isProcessing: boolean = false;
  private processInterval: NodeJS.Timeout | null = null;
  
  start() {
    if (this.processInterval) {
      logger.warn('Message queue processor already running');
      return;
    }
    
    logger.info('Starting message queue processor');
    this.processInterval = setInterval(() => {
      this.processQueue();
    }, config.queue.processIntervalMs);
    
    // Initial process
    this.processQueue();
  }
  
  stop() {
    if (this.processInterval) {
      clearInterval(this.processInterval);
      this.processInterval = null;
      logger.info('Message queue processor stopped');
    }
  }
  
  async addToQueue(
    clientId: string,
    phoneNumber: string,
    content: string
  ): Promise<string> {
    // Create message record
    const message = await prisma.message.create({
      data: {
        clientId,
        phoneNumber,
        content,
        status: MessageStatus.QUEUED,
      },
    });
    
    // Add to queue
    await prisma.messageQueue.create({
      data: {
        messageId: message.id,
        priority: 0,
        maxAttempts: config.rateLimits.maxRetries,
        status: QueueStatus.PENDING,
      },
    });
    
    logger.info(`Message ${message.id} added to queue for ${phoneNumber}`);
    return message.id;
  }
  
  async addBulkToQueue(
    messages: Array<{ clientId: string; phoneNumber: string; content: string }>
  ): Promise<string[]> {
    const messageIds: string[] = [];
    
    for (const msg of messages) {
      const messageId = await this.addToQueue(msg.clientId, msg.phoneNumber, msg.content);
      messageIds.push(messageId);
    }
    
    return messageIds;
  }
  
  private async processQueue() {
    if (this.isProcessing) {
      return;
    }
    
    this.isProcessing = true;
    
    try {
      // Get pending messages from queue
      const queuedItems = await prisma.messageQueue.findMany({
        where: {
          status: QueueStatus.PENDING,
          nextRetry: {
            lte: new Date(),
          },
          attempts: {
            lt: prisma.messageQueue.fields.maxAttempts,
          },
        },
        orderBy: [
          { priority: 'desc' },
          { createdAt: 'asc' },
        ],
        take: config.queue.batchSize,
      });
      
      for (const queueItem of queuedItems) {
        await this.processQueueItem(queueItem.id, queueItem.messageId);
      }
    } catch (error) {
      logger.error(`Error processing queue: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      this.isProcessing = false;
    }
  }
  
  private async processQueueItem(queueId: string, messageId: string) {
    try {
      // Update queue status
      await prisma.messageQueue.update({
        where: { id: queueId },
        data: { status: QueueStatus.PROCESSING },
      });
      
      // Get message details
      const message = await prisma.message.findUnique({
        where: { id: messageId },
      });
      
      if (!message) {
        logger.error(`Message ${messageId} not found`);
        await prisma.messageQueue.update({
          where: { id: queueId },
          data: { 
            status: QueueStatus.FAILED,
            lastError: 'Message not found',
          },
        });
        return;
      }
      
      // Update message status
      await prisma.message.update({
        where: { id: messageId },
        data: { status: MessageStatus.SENDING },
      });
      
      // Send SMS
      const result = await smsService.sendSms(message.phoneNumber, message.content);
      
      if (result.success) {
        // Success - mark as completed
        await prisma.message.update({
          where: { id: messageId },
          data: {
            status: MessageStatus.SENT,
            sentAt: new Date(),
          },
        });
        
        await prisma.messageQueue.update({
          where: { id: queueId },
          data: { status: QueueStatus.COMPLETED },
        });
        
        logger.info(`Message ${messageId} sent successfully`);
      } else if (result.retryable) {
        // Retryable error
        const queueItem = await prisma.messageQueue.findUnique({
          where: { id: queueId },
        });
        
        const newAttempts = (queueItem?.attempts || 0) + 1;
        const isDeadLetter = newAttempts >= config.rateLimits.maxRetries;
        
        await prisma.messageQueue.update({
          where: { id: queueId },
          data: {
            status: isDeadLetter ? QueueStatus.DEAD_LETTER : QueueStatus.PENDING,
            attempts: newAttempts,
            lastError: result.error,
            nextRetry: new Date(Date.now() + config.rateLimits.retryDelayMs * Math.pow(2, newAttempts - 1)),
          },
        });
        
        await prisma.message.update({
          where: { id: messageId },
          data: {
            status: isDeadLetter ? MessageStatus.FAILED : MessageStatus.QUEUED,
            retryCount: newAttempts,
            errorMessage: result.error,
          },
        });
        
        logger.warn(`Message ${messageId} will be retried (attempt ${newAttempts})`);
      } else {
        // Non-retryable error
        await prisma.message.update({
          where: { id: messageId },
          data: {
            status: MessageStatus.FAILED,
            errorMessage: result.error,
          },
        });
        
        await prisma.messageQueue.update({
          where: { id: queueId },
          data: {
            status: QueueStatus.FAILED,
            lastError: result.error,
          },
        });
        
        logger.error(`Message ${messageId} failed permanently: ${result.error}`);
      }
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      logger.error(`Error processing queue item ${queueId}: ${errorMsg}`);
      
      await prisma.messageQueue.update({
        where: { id: queueId },
        data: {
          status: QueueStatus.PENDING,
          lastError: errorMsg,
          nextRetry: new Date(Date.now() + config.rateLimits.retryDelayMs),
        },
      });
    }
  }
  
  async getQueueStats(): Promise<{
    pending: number;
    processing: number;
    completed: number;
    failed: number;
    deadLetter: number;
  }> {
    const [pending, processing, completed, failed, deadLetter] = await Promise.all([
      prisma.messageQueue.count({ where: { status: QueueStatus.PENDING } }),
      prisma.messageQueue.count({ where: { status: QueueStatus.PROCESSING } }),
      prisma.messageQueue.count({ where: { status: QueueStatus.COMPLETED } }),
      prisma.messageQueue.count({ where: { status: QueueStatus.FAILED } }),
      prisma.messageQueue.count({ where: { status: QueueStatus.DEAD_LETTER } }),
    ]);
    
    return { pending, processing, completed, failed, deadLetter };
  }
  
  async retryDeadLetterMessages(): Promise<number> {
    const result = await prisma.messageQueue.updateMany({
      where: { status: QueueStatus.DEAD_LETTER },
      data: {
        status: QueueStatus.PENDING,
        attempts: 0,
        nextRetry: new Date(),
      },
    });
    
    logger.info(`Reset ${result.count} dead letter messages for retry`);
    return result.count;
  }
}

export const messageQueueService = new MessageQueueService();

File: backend/src/main/java/com/clientmanagement/service/MessageQueueService.java
package com.clientmanagement.service;

import com.clientmanagement.config.QueueConfig;
import com.clientmanagement.config.SmsConfig;
import com.clientmanagement.dto.QueueStatusDto;
import com.clientmanagement.entity.Client;
import com.clientmanagement.entity.Message;
import com.clientmanagement.entity.MessageQueue;
import com.clientmanagement.entity.MessageStatus;
import com.clientmanagement.entity.QueueStatus;
import io.quarkus.scheduler.Scheduled;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import org.jboss.logging.Logger;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Service for managing the SMS message queue.
 */
@ApplicationScoped
public class MessageQueueService {

    private static final Logger LOG = Logger.getLogger(MessageQueueService.class);

    @Inject
    SmsService smsService;

    @Inject
    SmsConfig smsConfig;

    @Inject
    QueueConfig queueConfig;

    private volatile boolean isProcessing = false;

    /**
     * Add a message to the queue.
     */
    @Transactional
    public String addToQueue(String clientId, String phoneNumber, String content) {
        // Create message record
        Message message = new Message();
        message.id = UUID.randomUUID().toString();
        message.client = Client.findById(clientId);
        message.phoneNumber = phoneNumber;
        message.content = content;
        message.status = MessageStatus.QUEUED;
        message.retryCount = 0;
        message.createdAt = LocalDateTime.now();
        message.updatedAt = LocalDateTime.now();
        message.persist();

        // Create queue entry
        MessageQueue queueItem = new MessageQueue();
        queueItem.id = UUID.randomUUID().toString();
        queueItem.messageId = message.id;
        queueItem.priority = 0;
        queueItem.attempts = 0;
        queueItem.maxAttempts = smsConfig.rateLimit().maxRetries();
        queueItem.nextRetry = LocalDateTime.now();
        queueItem.status = QueueStatus.PENDING;
        queueItem.createdAt = LocalDateTime.now();
        queueItem.updatedAt = LocalDateTime.now();
        queueItem.persist();

        LOG.infof("Message %s added to queue for %s", message.id, phoneNumber);
        return message.id;
    }

    /**
     * Add multiple messages to the queue.
     */
    @Transactional
    public List<String> addBulkToQueue(List<BulkMessageItem> messages) {
        List<String> messageIds = new ArrayList<>();
        for (BulkMessageItem msg : messages) {
            String messageId = addToQueue(msg.clientId, msg.phoneNumber, msg.content);
            messageIds.add(messageId);
        }
        return messageIds;
    }

    public static class BulkMessageItem {
        public String clientId;
        public String phoneNumber;
        public String content;

        public BulkMessageItem(String clientId, String phoneNumber, String content) {
            this.clientId = clientId;
            this.phoneNumber = phoneNumber;
            this.content = content;
        }
    }

    /**
     * Process the message queue (scheduled task).
     */
    @Scheduled(every = "${queue.process-interval-seconds:5}s")
    @Transactional
    public void processQueue() {
        if (isProcessing) {
            return;
        }

        isProcessing = true;
        try {
            List<MessageQueue> queuedItems = MessageQueue.find(
                "status = ?1 AND nextRetry <= ?2 AND attempts < ?3 ORDER BY priority DESC, createdAt ASC",
                QueueStatus.PENDING,
                LocalDateTime.now(),
                smsConfig.rateLimit().maxRetries()
            ).page(0, queueConfig.batchSize()).list();

            for (MessageQueue queueItem : queuedItems) {
                processQueueItem(queueItem);
            }
        } catch (Exception e) {
            LOG.errorf("Error processing queue: %s", e.getMessage());
        } finally {
            isProcessing = false;
        }
    }

    private void processQueueItem(MessageQueue queueItem) {
        try {
            // Update queue status
            queueItem.status = QueueStatus.PROCESSING;
            queueItem.updatedAt = LocalDateTime.now();
            queueItem.persist();

            // Get message details
            Message message = Message.findById(queueItem.messageId);
            if (message == null) {
                LOG.errorf("Message %s not found", queueItem.messageId);
                queueItem.status = QueueStatus.FAILED;
                queueItem.lastError = "Message not found";
                queueItem.updatedAt = LocalDateTime.now();
                queueItem.persist();
                return;
            }

            // Update message status
            message.status = MessageStatus.SENDING;
            message.updatedAt = LocalDateTime.now();
            message.persist();

            // Send SMS
            SmsService.SmsSendResult result = smsService.sendSms(message.phoneNumber, message.content);

            if (result.success) {
                // Success
                message.status = MessageStatus.SENT;
                message.sentAt = LocalDateTime.now();
                message.updatedAt = LocalDateTime.now();
                message.persist();

                queueItem.status = QueueStatus.COMPLETED;
                queueItem.updatedAt = LocalDateTime.now();
                queueItem.persist();

                LOG.infof("Message %s sent successfully", queueItem.messageId);
            } else if (result.retryable) {
                // Retryable error
                int newAttempts = queueItem.attempts + 1;
                boolean isDeadLetter = newAttempts >= smsConfig.rateLimit().maxRetries();
                long delayMs = smsConfig.rateLimit().retryDelayMs() * (long) Math.pow(2, newAttempts - 1);
                LocalDateTime nextRetry = LocalDateTime.now().plusNanos(delayMs * 1_000_000);

                queueItem.status = isDeadLetter ? QueueStatus.DEAD_LETTER : QueueStatus.PENDING;
                queueItem.attempts = newAttempts;
                queueItem.lastError = result.error;
                queueItem.nextRetry = nextRetry;
                queueItem.updatedAt = LocalDateTime.now();
                queueItem.persist();

                message.status = isDeadLetter ? MessageStatus.FAILED : MessageStatus.QUEUED;
                message.retryCount = newAttempts;
                message.errorMessage = result.error;
                message.updatedAt = LocalDateTime.now();
                message.persist();

                LOG.warnf("Message %s will be retried (attempt %d)", queueItem.messageId, newAttempts);
            } else {
                // Non-retryable error
                message.status = MessageStatus.FAILED;
                message.errorMessage = result.error;
                message.updatedAt = LocalDateTime.now();
                message.persist();

                queueItem.status = QueueStatus.FAILED;
                queueItem.lastError = result.error;
                queueItem.updatedAt = LocalDateTime.now();
                queueItem.persist();

                LOG.errorf("Message %s failed permanently: %s", queueItem.messageId, result.error);
            }
        } catch (Exception e) {
            LOG.errorf("Error processing queue item %s: %s", queueItem.id, e.getMessage());
            queueItem.status = QueueStatus.PENDING;
            queueItem.lastError = e.getMessage();
            queueItem.nextRetry = LocalDateTime.now().plusNanos(smsConfig.rateLimit().retryDelayMs() * 1_000_000);
            queueItem.updatedAt = LocalDateTime.now();
            queueItem.persist();
        }
    }

    /**
     * Get queue statistics.
     */
    public QueueStatusDto.QueueStats getQueueStats() {
        QueueStatusDto.QueueStats stats = new QueueStatusDto.QueueStats();
        stats.pending = (int) MessageQueue.count("status", QueueStatus.PENDING);
        stats.processing = (int) MessageQueue.count("status", QueueStatus.PROCESSING);
        stats.completed = (int) MessageQueue.count("status", QueueStatus.COMPLETED);
        stats.failed = (int) MessageQueue.count("status", QueueStatus.FAILED);
        stats.deadLetter = (int) MessageQueue.count("status", QueueStatus.DEAD_LETTER);
        return stats;
    }

    /**
     * Retry dead letter messages.
     */
    @Transactional
    public int retryDeadLetterMessages() {
        int count = MessageQueue.update(
            "status = ?1, attempts = 0, nextRetry = ?2, updatedAt = ?2 WHERE status = ?3",
            QueueStatus.PENDING,
            LocalDateTime.now(),
            QueueStatus.DEAD_LETTER
        );
        LOG.infof("Reset %d dead letter messages for retry", count);
        return count;
    }
}
